demos/common/monitors/cpu_monitor.cpp

```diff
@@ -0,0 +1,274 @@
+    lastEnabled = true;
+    if (!historyEnabled) {
+        historySize = 1;
+        openQuery();
+    }
+}
+
+bool CpuMonitor::isLastEnabled() const {
+    return lastEnabled;
+}
+
+void CpuMonitor::disableLast() {
+    lastEnabled = false;
+    if (!historyEnabled) {
+        closeQuery();
+    }
+}
+
+void CpuMonitor::collectData() {
+    // store CPU Utulization in range [0, 1]
+    PDH_STATUS status;
+    PDH_FMT_COUNTERVALUE DisplayValue;
+
+    status = PdhCollectQueryData(query);
+    if (ERROR_SUCCESS != status) {
+        throw std::runtime_error("PdhCollectQueryData() failed");
+    }
+
+    std::vector<std::pair<double, double>> idleNonIdleCpuStat;
+    for (std::size_t i = 0; i < coreTimeCounters.size(); i++) {
+        status = PdhGetFormattedCounterValue(coreTimeCounters[i], PDH_FMT_DOUBLE, NULL, &DisplayValue);
+        if (ERROR_SUCCESS != status) {
+            throw std::runtime_error("PdhGetFormattedCounterValue() failed");
+        }
+
+        idleNonIdleCpuStat.emplace_back(DisplayValue.doubleValue, 1 - DisplayValue.doubleValue);
+    }
+
+    std::vector<double> cpuLoad(idleNonIdleCpuStat.size());
+    for (std::size_t i = 0; i < idleNonIdleCpuStat.size(); ++i) {
+        cpuLoad[i] = static_cast<double>(idleNonIdleCpuStat[i].first);
+    }
+
+    if (meanCpuLoad.empty()) {
+        meanCpuLoad = cpuLoad;
+        samplesNumber = 1;
+    } else {
+        for (std::size_t i = 0; i < meanCpuLoad.size(); ++i) {
+            meanCpuLoad[i] = (meanCpuLoad[i] * samplesNumber + cpuLoad[i]) / (samplesNumber + 1);
+        }
+        ++samplesNumber;
+    }
+    cpuLoadHistory.push_back(std::move(cpuLoad));
+    if (cpuLoadHistory.size() > historySize) {
+        cpuLoadHistory.pop_front();
+    }
+}
+
+std::deque<std::vector<double>> CpuMonitor::getLastHistory() const {
+    return cpuLoadHistory;
+}
+
+std::vector<double> CpuMonitor::getMeanCpuLoad() const {
+    return meanCpuLoad;
+}
+#else
+#include <regex>
+#include <utility>
+#include <fstream>
+
+namespace {
+std::size_t getNCores() {
+    std::ifstream cpuinfo("/proc/cpuinfo");
+    return std::count(std::istream_iterator<std::string>(cpuinfo),
+        std::istream_iterator<std::string>(),
+        std::string("processor"));
+}
+
+std::vector<std::pair<unsigned long, unsigned long>> getIdleNonIdleCpuStat(std::size_t nCores) {
+    std::vector<std::pair<unsigned long, unsigned long>> idleNonIdleCpuStat(nCores);
+    std::ifstream proc_stat("/proc/stat");
+    std::string line;
+    std::smatch match;
+    std::getline(proc_stat, line);
+    while (proc_stat.good())
+    {
+        std::regex core_jiffies("^cpu(\\d+)\\s+"
+            "(\\d+)\\s+" // user
+            "(\\d+)\\s+" // nice
+            "(\\d+)\\s+" // system
+            "(\\d+)\\s+" // idle
+            "(\\d+)\\s+" // iowait
+            "(\\d+)\\s+" // irq
+            "(\\d+)\\s+" // softirq
+            "(\\d+)\\s+" // steal
+            "(\\d+)\\s+" // guest
+            "(\\d+)$");  // guest_nice
```

Comment:
I wouldn't recommend anchoring the regex at the end, since Linux might add new columns in the future.

Reply:
This will us to now, that `/proc/stat` is different.

Reply:
Why would we want to know?

Reply:
To update regex.

Reply:
That's true, I guess, but the code will be broken until we do. I think it would be better to assume that any additional fields won't affect our calculations, and just ignore them.

Reply:
updated